### PR TITLE
doc: adds 'How function call works'

### PR DESCRIPTION
**Pull request by mathetake**

### CODE REVIEW

**Comment by mathetake** [comment]
@evacchi this is something you might be interested in 😎

**Comment by codefromthecrypt** [comment]
https://deploy-preview-1216--wazero.netlify.app/docs/how_function_call_works/

**Comment by codefromthecrypt** [comment]
renamed the doc on compiler functions as it isn't about the interpreter. also added more in the line of overview. Please correct if I did something wrong :D I'll review more later

**Comment by codefromthecrypt** [comment]
added another commit to reduce duplication in the initial bullets and also not introduce too many concepts. For example, I think the file cache isn't required to be introduced in order to understand the different engine types.

**Comment by mathetake** [comment]
Looking good

File: site/content/docs/how_do_compiler_functions_work.md [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+example. `clock_time_get`, is a host function defined in Go, called from
+machine code compiled from guest wasm. Let's say the wasm function is named
+`getTime`.
+
+When `getTime` calls  `clock_time_get`, it actually exits execution first.
+wazero then calls the Go function mapped to `clock_time_get` like a usual Go
+program. Finally, wazero transfers control back to machine code again, resuming
+execution after the `clock_time_get` call instruction.
+
+TODO: add a diagram
```
Comment (codefromthecrypt) [inline site/content/docs/how_do_compiler_functions_work.md:81]: @mathetake can I trouble you with a diagram, using the function names above, not foo bar?

Comment (mathetake) [reply site/content/docs/how_do_compiler_functions_work.md:81]: the diagram in the next section is exactly for this.

Comment (codefromthecrypt) [reply site/content/docs/how_do_compiler_functions_work.md:81]: top-to-bottom editing problem :D

**Comment by codefromthecrypt** [comment]
@mathetake PTAL! https://deploy-preview-1216--wazero.netlify.app/docs/how_do_compiler_functions_work/

**Comment by mathetake** [comment]
coolio

**Pull request merged**